source/common/router/scoped_config_impl.cc

```diff
@@ -3,6 +3,107 @@
 namespace Envoy {
 namespace Router {
 
+bool ScopeKey::operator!=(const ScopeKey& other) const { return !(*this == other); }
+
+bool ScopeKey::operator==(const ScopeKey& other) const {
+  if (this->fragments_.empty() || other.fragments_.empty()) {
+    // An empty key equals to nothing, "NULL" != "NULL".
+    return false;
+  }
+  return std::equal(this->fragments_.begin(), this->fragments_.end(), other.fragments_.begin(),
+                    other.fragments_.end(),
+                    [](const std::unique_ptr<ScopeKeyFragmentBase>& left,
+                       const std::unique_ptr<ScopeKeyFragmentBase>& right) -> bool {
+                      // Both should be non-NULL now.
+                      return *left == *right;
+                    });
+}
+
+HeaderValueExtractorImpl::HeaderValueExtractorImpl(
+    ScopedRoutes::ScopeKeyBuilder::FragmentBuilder config)
+    : FragmentBuilderBase(std::move(config)),
+      header_value_extractor_config_(config_.header_value_extractor()) {
+  ASSERT(config_.type_case() ==
+             ScopedRoutes::ScopeKeyBuilder::FragmentBuilder::kHeaderValueExtractor,
+         "header_value_extractor is not set.");
+  if (header_value_extractor_config_.extract_type_case() ==
+      ScopedRoutes::ScopeKeyBuilder::FragmentBuilder::HeaderValueExtractor::kIndex) {
+    if (header_value_extractor_config_.index() != 0 &&
+        header_value_extractor_config_.element_separator().length() == 0) {
+      throw ProtoValidationException("when element separator is set to an empty string, index "
```

Comment:
Nit: prefer descriptive rather than didactic exception messages, e.g. "Index > 0 for empty string element separator". The idea is we're describing a failure condition, not providing a help system :)

Reply:
right, C&P.